aws/private/s3_sync.sh

```diff
@@ -0,0 +1,219 @@
+#!/usr/bin/env bash
+#
+# Copyright 2022 Aspect Build Systems, Inc.
+#
+# s3 cp/sync script intended for use with s3_sync rule
+
+set -o errexit -o nounset -o pipefail
+
+# Functions
+
+msg() {
+    if [[ $# -gt 0 ]]; then
+        local msg="${1:-}"
+        shift 1
+        while (("$#")); do
+            msg="${msg:-}"$'\n'"${1}"
+            shift 1
+        done
+        echo "${msg}"
+    else
+        cat
+    fi
+}
+
+warn() {
+    if [[ $# -gt 0 ]]; then
+        local msg="${1:-}"
+        shift 1
+        while (("$#")); do
+            msg="${msg:-}"$'\n'"${1}"
+            shift 1
+        done
+        echo >&2 "${msg}"
+    else
+        cat >&2
+    fi
+}
+
+# Echos the provided message to stderr and exits with an error (1).
+# shellcheck disable=SC2120
+fail() {
+    warn "$@"
+    exit 1
+}
+
+# Print an error message and dump the usage/help for the utility.
+# This function expects a get_usage function to be defined.
+usage_error() {
+    local msg="${1:-}"
+    cmd=(fail)
+    [[ -z "${msg:-}" ]] || cmd+=("${msg}" "")
+    cmd+=("$(get_usage)")
+    "${cmd[@]}"
+}
+
+show_usage() {
+    get_usage
+    exit 0
+}
+
+get_usage() {
+    local utility
+    utility="$(basename "${BASH_SOURCE[0]}")"
+    cat <<-'EOF'
+Copies the specified artifacts to an S3 bucket.
+
+Usage:
+EOF
+    echo "${utility} [OPTION]... <artifact>..."
+    cat <<-'EOF'
+Options:
+  --bucket                The name of the S3 bucket.
+  --bucket_file <file>    The path to a file that contains the name of the S3 bucket.
+  --[no]dry_run           Toggles whether the utility will run in dry-run mode.
+                          Default: false
+  --sync_archive          Rather than copy an archive file, unpack it and sync the contents.
+
+Arguments:
+  <artifact>              The path to a file which will be copied to the S3 bucket.
+                          One or more artifacts can be specified.
+EOF
+}
+
+cp_artifact() {
+    local artifact="${1}"
+    local bucket="${2}"
+
+    if [ -d "${artifact}" ]; then
+        # Always flatten directories
+        for f in "${artifact}"/*; do
+            cp_artifact "${f}" "${bucket}"
+        done
+    elif [[ "${sync_archive:-}" ]]; then
+        if [[ "${dry_run}" == "false" ]]; then
+            warn "Syncing ${artifact} to ${bucket}"
+            local untar
+            untar=$(mktemp -d)
+            tar -xf "${artifact}" -C "${untar}"
```

Comment:
use bsdtar toolchain?

Reply:
I'll remove the `--sync_archive` option for this PR, need to land for a client and not sure if they have that requirement. Easy to add to the API in a follow-up.